### DISCUSSION TITLE
AHCI: Page fault when enumerating HBA memory space

### DISCUSSION THREAD

**Discussion by kennystrawnmusic** [phil-opp/blog_os#1174] [Questions] [answered]
Can anyone explain why [this](https://github.com/kennystrawnmusic/cryptos/blob/7bee5ac28930f7ae1633a822ad01b2b7b9797704/src/ahci/mod.rs#L50) would cause a page fault by overflowing the stack?

Was able to narrow the source down to precisely that stage in the AHCI initialization process by attempting to print test lines onto the screen and seeing which point those test lines would no longer print due to the page fault panic. The point at which the test lines no longer print is at the very top of the "filter_map" closure.

The only possibility I can think of is the use of heap-allocated structures (`Vec<Box<dyn Disk + Send + Sync>>`) to point to stack-allocated entries in HBA memory, but that makes no sense given that Redox manages to succeed at exactly that.

**Comment by kennystrawnmusic** [comment]
To further elaborate: it appears to be some sort of overflow. Attempts to access out-of-bounds memory addresses and when I go on to allocate a page at each offending address, it attempts to access even more out-of-bounds memory each and every time. Which is strange seeing as though creating a vector of disks shouldn't cause a stack overflow, and neither should what is essentially a clone of the Redox AHCI driver with kernel-mode-compatibility tweaking.

**Reply by kennystrawnmusic** [reply]
Since @bjorn3 is the only one who's ever answered anything before, getting his attention. [This](https://github.com/kennystrawnmusic/cryptos/commit/7bee5ac28930f7ae1633a822ad01b2b7b9797704) is all that I've attempted, and none of that is fixing the problem.

**Reply by bjorn3** [reply]
I looked at your code and couldn't find anything obviously wrong. I'm not too well versed with ACPI and don't know anything about AHCI at all, so I don't think I can help you. You said you ported this code from Redox OS, right? Maybe you could try asking at https://chat.redox-os.org/?

**Reply by kennystrawnmusic** [reply]
> You said you ported this code from Redox OS, right? Maybe you could try asking at https://chat.redox-os.org/?

I spent weeks studying [this code](https://gitlab.redox-os.org/redox-os/drivers/-/tree/master/ahcid/src/ahci) as a way to attempt to understand how AHCI drivers are written generally, then went on to write it from scratch using the information gathered from that case study.

There are a several key differences — notably using [my own `Mmio` struct](https://github.com/kennystrawnmusic/cryptos/blob/7bee5ac28930f7ae1633a822ad01b2b7b9797704/src/ahci/hba/structs.rs#L17) instead of relying on `redox_syscall`'s, [splitting the `hba` module into three separate source files](https://github.com/kennystrawnmusic/cryptos/tree/master/src/ahci/hba) to clean it up a bit, and giving the struct fields far more human-readable names than the ones Redox uses, for starters. Then there's the fact that [this code](https://gitlab.redox-os.org/redox-os/drivers/-/blob/master/ahcid/src/ahci/hba.rs#L161) is DRY, so I used [a macro](https://github.com/kennystrawnmusic/cryptos/blob/7bee5ac28930f7ae1633a822ad01b2b7b9797704/src/ahci/hba/mod.rs#L38) to clean it up. Moreover, careful reading of Intel's AHCI documentation allowed me to create [an enum](https://github.com/kennystrawnmusic/cryptos/blob/7bee5ac28930f7ae1633a822ad01b2b7b9797704/src/ahci/hba/structs.rs#L229) to handle I/O errors in a far more user-friendly manner than the way Redox does, which is by merely printing the error codes instead of matching human-readable strings to them as I do. Finally, my implementation doesn't have a `Dma` structure at all because you don't need one if the driver is baked into the kernel; Redox needs one because their ahcid is a user-mode driver. Despite these key differences, however, the overall design idea is similar and could potentially end up being superior to the design that Redox settled on — if I can get it to stop smashing the kernel stack (a misbehavior that has wasted months of my time with no progress which is exactly why I posted here), that is.

I wonder if @IsaacWoods can help figure this out seeing as though he has a kernel written that is similar to Redox's in a sense that it's also a microkernel. Mine is supposed to eventually be hybrid, however, so maybe not.

**Reply by IsaacWoods** [reply]
> Since @bjorn3 is the only one who's ever answered anything before, getting his attention. [This](https://github.com/kennystrawnmusic/cryptos/commit/7bee5ac28930f7ae1633a822ad01b2b7b9797704) is all that I've attempted, and none of that is fixing the problem.

I seem to remember answering plenty of ACPI questions before...

> Can anyone explain why [this](https://github.com/kennystrawnmusic/cryptos/blob/7bee5ac28930f7ae1633a822ad01b2b7b9797704/src/ahci/mod.rs#L50) would cause a page fault by overflowing the stack?

This is a bit of a stretch, but my first guess would always be to increase the amount of stack Rust has - `debug`-compiled code can use an eye-watering amount of stack (I've literally seen it use megabytes with badly-written code). I don't know how much stack `bootloader` gives you, but my first try would be increasing it, and then increasing it some more.

I'd also build a mechanism to reliably detect kernel stack overflow in your page-fault handler (generally with guard pages), and report it better so it doesn't seem like 'random' page-faults.

**Reply by kennystrawnmusic** [reply]
>This is a bit of a stretch, but my first guess would always be to increase the amount of stack Rust has

Attempting to increase the stack size causes the bootloader to panic with a "frame allocation failed when mapping a kernel stack" message when a 128MB kernel stack is attempted and even a 64MB kernel stack doesn't do anything to fix this. It's also compiled by default with `--release` (and custom settings to bring back certain symbols) to begin with, yet still smashing the kernel stack despite that.

**Reply by kennystrawnmusic** [reply]
> I'd also build a mechanism to reliably detect kernel stack overflow in your page-fault handler (generally with guard pages), and report it better so it doesn't seem like 'random' page-faults.

Does the bootloader provide an API for reliably detecting where to place a guard page in memory?

**Reply by kennystrawnmusic** [reply]
For further information, the address that the driver is trying to access only to trigger a page fault is 0xfffffff720. Whenever I try to map a page there, I get the same page fault at another higher address, and no matter how many pages I map, the same problem persists. Something is infinitely recursing in the code of that driver and I'm still trying to figure out where.

**Reply by kennystrawnmusic** [reply]
Here's the output thus far. You'll notice that every time I try to allocate a page at the address it faults at, it keeps faulting at a lower and lower address:
<img width="1440" alt="Screenshot 2022-11-17 at 10 27 28 PM" src="https://user-images.githubusercontent.com/3030773/202635922-9036858a-c447-4e18-9d47-782c099c29a7.png">
<img width="1440" alt="Screenshot 2022-11-17 at 10 34 33 PM" src="https://user-images.githubusercontent.com/3030773/202638293-a94a15dc-9eb9-4f30-a9f6-0ebd4eec846e.png">
<img width="1440" alt="Screenshot 2022-11-17 at 10 42 51 PM" src="https://user-images.githubusercontent.com/3030773/202638822-e399b96f-577f-496e-a962-d83b8e20fd42.png">


And when I attempt to put a huge page there:
<img width="1440" alt="Screenshot 2022-11-17 at 10 39 54 PM" src="https://user-images.githubusercontent.com/3030773/202638351-6aabe2d1-bf56-4bfd-8f40-8629b76d56fe.png">


So something is clearly recursing. Still struggling after months of this to figure out exactly what.

**Reply by kennystrawnmusic** [reply]
Managed to at least get it to detect a disk by [moving the start of the kernel stack to the extreme high-end](https://github.com/kennystrawnmusic/cryptos/commit/4da4ae6eacf5b11b45c9824ec835193c629f1104#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fc). Page fault still persists however; it just happens after detection of the first disk instead of before.

**Reply by kennystrawnmusic** [reply]
After a lengthy struggle attempting to find a GPLv3 replacement, finally managed to in Aero's AHCI implementation, so decided to port that instead of attempting to reinvent the wheel by rewriting the AHCI driver that comes with Redox. Page fault is now gone, but I've run into a new problem:
![image](https://user-images.githubusercontent.com/3030773/207777546-09053e96-a6e3-4981-9d40-747bbd54258f.jpeg)

The IDT is only supposed to be 256 entries long, no? So why would a #NP fault expecting a 302nd entry in the IDT even be possible? Or is there something else going on here?

**Comment by kennystrawnmusic** [comment]
Nearly a whole month and an attempted complete rewrite of the AHCI driver later (by porting Aero's implementation instead of trying to rewrite Redox's since Aero is GPLv3 and so is my own kernel, so no licensing issues) I finally have a promising lead: Turns out QEMU is placing the ABAR in legacy I/O port space instead of MMIO space. The hilarious thing about this is I had to comment out the assertion check of the header type because for some reason it isn't seeing a zero (normal) there despite the fact that the pcics crate *is* correctly seeing it as a normal header.

Is there any way to prevent this and/or force-enable MMIO disk access from the QEMU side of things?

**Comment by kennystrawnmusic** [comment, marked as answer]
Finally figured out the problem after taking to StackOverflow for help: the problem is rust-osdev/x86_64#396. `SelectorErrorCode::index()` is performing a misaligned read of the bits in the selector code given by the fault.
